### DISCUSSION TITLE
Double Faults (zh-CN)

### DISCUSSION THREAD

**Discussion by giscus** [phil-opp/blog_os#1214] [Post Comments (translated)]
# Double Faults (zh-CN)

在这篇文章中，我们会探索 double fault 异常的细节，它的触发条件是调用错误处理函数失败。通过捕获该异常，我们可以阻止致命的 triple faults 异常导致系统重启。为了尽可能避免 triple faults ，我们会在一个独立的内核栈配置 中断栈表 来捕捉 double fault…

https://os.phil-opp.com/zh-CN/double-fault-exceptions/

**Comment by nliao6622** [comment]
Well, first, this is a very good tutorial. 

A quick question: when I try to repeat the page fault at 0xdeadbeef in very first example on macbook, I got this following message:

"panic at 'misaligned pointer dereference: address must be a multiple of 0x8 but is 0xdeadbeef' src/main.rs:21:9"

More interestingly, the fault is the same even after I defined the double_fault handler. Looks the double_fault_handler is never triggered. 

My platform is Macbook Big Sur Intel based. What could be the like cause of the problem? 

Thanks!

**Reply by ghost** [reply]
I had the same issue. The address has to be a multiple of 8, which 0xdeadbeef (3735928559 in decimal)
is not. Try using 0xdeadbef0 (3735928560). This might be helpful http://junyelee.blogspot.com/2019/01/memory-alignment-issue.html

**Reply by SPuntte** [reply]
I stumbled upon this as well. Indeed, the address `0xdeadbeef` is not properly aligned for a `*mut u64`. Rust considers such dereferences [undefined behavior](https://doc.rust-lang.org/reference/behavior-considered-undefined.html), and the compiler [recently received the facilities to generate checks](https://github.com/rust-lang/rust/pull/98112) for this. Thus the unexpected panic message.

Another option would be to substitute `u8` for `u64`:
```Rust
// trigger a page fault
unsafe {
    *(0xdeadbeef as *mut u8) = 42;
};
```

**Reply by nliao6622** [reply]
Yes using the "0xdeadbee0" definitely helps.

**Reply by phil-opp** [reply]
This should be fixed by https://github.com/phil-opp/blog_os/pull/1226.

**Comment by SPuntte** [comment]
As a meta question, surely the English-language blog post is not supposed to show this "zh-CN" discussion thread?

**Reply by phil-opp** [reply]
Thanks for reporting this! There was an issue in the comment set up that I just fixed in #1230. I also moved the existing discussions into the correct categories.

**Comment by allen7593** [comment]
Hi,
Great work on this blog and I am really enjoy learning it, bug I am having an issue with setting double fault handler, not sure why, I am new to rust and os, can anyone help me with it? Thanks in advance.

when I follow this blog I am having this error
```
error: invalid signature for `extern "x86-interrupt"` function                                                                                                                                                                     
  --> src\interrupts.rs:23:60
   |
23 |     stack_frame: InterruptStackFrame, _error_code: u64) -> !
   |                                                            ^
   |
   = note: functions with the "custom" ABI cannot have a return type
help: remove the return type
  --> src\interrupts.rs:23:60
   |
23 |     stack_frame: InterruptStackFrame, _error_code: u64) -> !
   |                                                            ^

error: could not compile `blog_os` (lib) due to 1 previous error  
```
it's telling me I should not have -> ! as a return type, but when I removed the -> !, it comes with another error:
```
error[E0308]: mismatched types                                                                                                                                                                                                     
   --> src\interrupts.rs:9:41
    |
9   |         idt.double_fault.set_handler_fn(double_fault_handler);
    |                          -------------- ^^^^^^^^^^^^^^^^^^^^ expected fn pointer, found fn item
    |                          |
    |                          arguments to this method are incorrect
    |
    = note: expected fn pointer `extern "x86-interrupt" fn(InterruptStackFrame, _) -> !`
                  found fn item `extern "x86-interrupt" fn(InterruptStackFrame, _) -> () {double_fault_handler}`
note: method defined here
   --> C:\Users\Allen Chen\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\x86_64-0.14.13\src\structures\idt.rs:795:12
    |
795 |     pub fn set_handler_fn(&mut self, handler: F) -> &mut EntryOptions {
    |            ^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
error: could not compile `blog_os` (lib) due to 1 previous error
```
It's telling me the function signature not comply with the design. 
I am a bit confused, I am suspecting this is a language issue or API changed?

here is the list of my env:
```
OS:: windows 11 24H2
Arch: x86_64
Cargo version: cargo 1.90.0-nightly (930b4f62c 2025-06-28)
rustc version: rustc 1.90.0-nightly (667787527 2025-07-02)
language edition: 2018
```

**Reply by EveGlowLuna** [reply]
Hi, I’ve actually run into this exact issue before, so I know how frustrating it can be. And good news — it’s not your code. This is a known regression in recent Rust nightly builds.

Starting from `nightly-2025-06-26`, there's a bug where the compiler incorrectly rejects the `-> !` return type for `extern "x86-interrupt"` functions. It first tells you to remove it, saying the function can't have a return type — but once you do, it complains that the signature is wrong because it *expects* `-> !`. You end up in this impossible loop: add the return type, it errors; remove it, it also errors. Yeah, it’s a real compiler bug. This has already been reported (see [rust-lang/rust#143072](https://github.com/rust-lang/rust/issues/143072) and [rust-lang/rust#143335](https://github.com/rust-lang/rust/issues/143335)). The same code compiles just fine on earlier versions like `nightly-2025-06-24`.

But wait — there’s another issue piling on top.  
The `x86_64` crate has some version compatibility problems too. The `Step` trait’s return type changed in newer Rust versions: from `Option<usize>` to `(usize, Option<usize>)`. So if your `x86_64` version doesn’t match your toolchain, you’ll hit errors like `steps_between` having the wrong return type. From what the community’s found, if you’re on a newer nightly (like 1.88+), you’ll probably need `x86_64 = "0.15.2"` to make things work. Older versions like `0.14.x` might only play nicely with older compilers.

To get unblocked quickly, the easiest fix is to roll back to a working nightly. Try running:

```bash
rustup override set nightly-2025-06-24
```

That should let you compile without hitting the `-> !` issue.

Hope that helps.

P.S. This was translated — I'm Chinese and still improving my English. Thanks so much for your patience and understanding!
